src/ops/shapes.rs

```diff
@@ -0,0 +1,2528 @@
+    arrays: impl IntoIterator<Item = impl AsRef<Array>>,
+    axis: i32,
+    stream: StreamOrDevice,
+) -> Array {
+    let arrays = arrays.into_iter().collect::<Vec<_>>();
+    stack_inner(&arrays, axis, stream)
+}
+
+#[default_device]
+pub fn try_stack_device(
+    arrays: impl IntoIterator<Item = impl AsRef<Array>>,
+    axis: i32,
+    stream: StreamOrDevice,
+) -> Result<Array, StackError> {
+    let arrays = arrays.into_iter().collect::<Vec<_>>();
+
+    if arrays.is_empty() {
+        return Err(StackError::NoInputArray);
+    }
+
+    if !is_same_shape(&arrays) {
+        return Err(StackError::InvalidShapes);
+    }
+
+    Ok(stack_inner(&arrays, axis, stream))
+}
+
+/// Stacks the arrays along a new axis. Panics if the arrays have different shapes.
+///
+/// # Params
+///
+/// - `arrays`: The input arrays.
+/// - `axis`: The axis in the result array along which the input arrays are stacked.
+///
+/// # Panics
+///
+/// Panics if the arrays have different shapes. See [`try_stack`] for more information.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let a = Array::from_iter(0..4, &[2, 2]);
+/// let b = Array::from_iter(4..8, &[2, 2]);
+/// let result = stack(&[&a, &b], 0);
+/// ```
+#[default_device]
+pub fn stack_device(
+    arrays: impl IntoIterator<Item = impl AsRef<Array>>,
+    axis: i32,
+    stream: StreamOrDevice,
+) -> Array {
+    try_stack_device(arrays, axis, stream).unwrap()
+}
+
+fn stack_all_inner(arrays: &[impl AsRef<Array>], stream: StreamOrDevice) -> Array {
+    unsafe {
+        let c_arrays = new_mlx_vector_array(arrays);
+        let c_array = mlx_sys::mlx_stack_all(c_arrays, stream.as_ptr());
+
+        let result = Array::from_ptr(c_array);
+        mlx_sys::mlx_free(c_arrays as *mut c_void);
+
+        result
+    }
+}
+
+/// Stacks the arrays along a new axis.
+///
+/// # Params
+///
+/// - `arrays`: The input arrays.
+///
+/// # Safety
+///
+/// The function is unsafe because it does not check if the arrays have the same shape.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let a = Array::from_iter(0..4, &[2, 2]);
+/// let b = Array::from_iter(4..8, &[2, 2]);
+/// let result = unsafe { stack_all_unchecked(&[&a, &b]) };
+/// ```
+#[default_device]
+pub unsafe fn stack_all_device_unchecked(
+    arrays: impl IntoIterator<Item = impl AsRef<Array>>,
+    stream: StreamOrDevice,
+) -> Array {
+    let arrays = arrays.into_iter().collect::<Vec<_>>();
+    stack_all_inner(&arrays, stream)
+}
+
+/// Stacks the arrays along a new axis. Returns an error if the arrays have different shapes.
+///
+/// # Params
+///
+/// - `arrays`: The input arrays.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let a = Array::from_iter(0..4, &[2, 2]);
+/// let b = Array::from_iter(4..8, &[2, 2]);
+/// let result = try_stack_all(&[&a, &b]);
+/// ```
+#[default_device]
+pub fn try_stack_all_device(
```

Comment:
same as try_stack_device

Reply:
Addressed in ec661bf335722cc03b7c0da398d6e683b0810bc1